File: COLLEGE/ASSIGNMENT6/c/Q.2.c
#include <stdio.h>
#include <stdlib.h>

// Node structure definition for linked list
struct Node {
    int data;
    struct Node* next;
};

// Function to check if the stack is empty
int isEmpty(struct Node* top) {
    return top == NULL;
}

// Function to push an element onto the stack
void push(struct Node** top, int value) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    if (newNode == NULL) {
        printf("Stack overflow! Cannot push %d\n", value);
        return;
    }
    newNode->data = value;
    newNode->next = *top;
    *top = newNode;
    printf("Pushed %d onto the stack\n", value);
}

// Function to pop an element from the stack
int pop(struct Node** top) {
    if (isEmpty(*top)) {
        printf("Stack underflow! Cannot pop\n");
        return -1;  // Returning -1 to indicate error
    }
    struct Node* temp = *top;
    int poppedValue = temp->data;
    *top = (*top)->next;
    free(temp);
    return poppedValue;
}

// Function to peek at the top element of the stack
int peek(struct Node* top) {
    if (isEmpty(top)) {
        printf("Stack is empty! No top element\n");
        return -1;  // Returning -1 to indicate error
    }
    return top->data;
}

// Main function to test the stack implementation
int main() {
    struct Node* stackTop = NULL;  // Initialize an empty stack

    push(&stackTop, 10);
    push(&stackTop, 20);
    push(&stackTop, 30);
    push(&stackTop, 40);

    printf("Top element is %d\n", peek(stackTop));

    printf("Popped element is %d\n", pop(&stackTop));
    printf("Popped element is %d\n", pop(&stackTop));

    push(&stackTop, 50);

    printf("Top element is %d\n", peek(stackTop));

    return 0;
}

File: COLLEGE/ASSIGNMENT6/python/Q.2.py
class Node:
    """A node of the linked list used in the stack."""
    def __init__(self, data):
        self.data = data
        self.next = None

class Stack:
    """A stack implementation using a linked list."""
    def __init__(self):
        self.top = None

    def is_empty(self):
        """Check if the stack is empty."""
        return self.top is None

    def push(self, item):
        """Add an item to the top of the stack."""
        new_node = Node(item)
        new_node.next = self.top
        self.top = new_node
        print(f"Pushed {item} onto stack.")

    def pop(self):
        """Remove and return the top item of the stack."""
        if self.is_empty():
            print("Stack Underflow! Cannot pop from an empty stack.")
            return None
        else:
            popped_node = self.top
            self.top = self.top.next
            print(f"Popped {popped_node.data} from stack.")
            return popped_node.data

    def peek(self):
        """Return the top item of the stack without removing it."""
        if self.is_empty():
            print("Stack is empty! Cannot peek.")
            return None
        else:
            return self.top.data

    def display(self):
        """Display the contents of the stack."""
        if self.is_empty():
            print("Stack is empty!")
            return
        current = self.top
        stack_contents = []
        while current:
            stack_contents.append(current.data)
            current = current.next
        print("Stack contents:", stack_contents)

# Example usage
if __name__ == "__main__":
    stack = Stack()

    stack.push(10)
    stack.push(20)
    stack.push(30)
    stack.display()

    print("Top element is:", stack.peek())
    
    stack.pop()
    stack.display()
    
    stack.pop()
    stack.pop()  # This should show underflow message

    stack.pop()  # This should show underflow message

    stack.push(40)
    stack.display()
